### PR TITLE
KAFKA-6122: Global Consumer should handle TimeoutException

### PR DESCRIPTION
**Pull request by mjsax**
Implements KIP-224:
- adding new StreamsConfig `retires`
- uses `retires` and `retry.backoff.ms` to handle TimeoutException in GlobalStateManager
- adds two new tests to trigger TimeoutException in global consumer
- some minor code cleanup to reduce number of parameters we need to pass around

### CODE REVIEW

**Comment by asfgit** [comment]
FAILURE
 6695 tests run, 1 skipped, 1 failed.
--none--

**Comment by asfgit** [comment]
SUCCESS 
 8069 tests run, 5 skipped, 0 failed.
--none--

**Comment by asfgit** [comment]
SUCCESS 
 8069 tests run, 5 skipped, 0 failed.
--none--

**Comment by mjsax** [comment]
Call for review @bbejeck @dguy @guozhangwang

**Comment by asfgit** [comment]
FAILURE
 6695 tests run, 1 skipped, 2 failed.
--none--

**Comment by asfgit** [comment]
FAILURE
 7961 tests run, 5 skipped, 1 failed.
--none--

**Comment by asfgit** [comment]
FAILURE
 7961 tests run, 5 skipped, 1 failed.
--none--

**Comment by mjsax** [comment]
Retest this please.

**Comment by asfgit** [comment]
FAILURE
 6695 tests run, 1 skipped, 2 failed.
--none--

**Comment by asfgit** [comment]
FAILURE
 6695 tests run, 1 skipped, 2 failed.
--none--

**Comment by asfgit** [comment]
FAILURE
 8069 tests run, 5 skipped, 2 failed.
--none--

**Comment by asfgit** [comment]
SUCCESS 
 8069 tests run, 5 skipped, 0 failed.
--none--

**Comment by asfgit** [comment]
FAILURE
 8069 tests run, 5 skipped, 2 failed.
--none--

**Comment by dguy** [comment]
@mjsax test failures maybe relevant? Same 2 tests failed in jdk7 & 8 builds

**Comment by mjsax** [comment]
@dguy Yes. Test passed locally though... But I guess it's related to setting state.dir to a proper test directory instead of default `/tmp/kafka-streams`. Updates this.

**Comment by asfgit** [comment]
SUCCESS 
 8069 tests run, 5 skipped, 0 failed.
--none--

**Comment by asfgit** [comment]
SUCCESS 
 8069 tests run, 5 skipped, 0 failed.
--none--

**Comment by asfgit** [comment]
SUCCESS 
 8069 tests run, 5 skipped, 0 failed.
--none--

**Comment by asfgit** [comment]
FAILURE
 4020 tests run, 1 skipped, 1 failed.
--none--

**Comment by asfgit** [comment]
FAILURE
 7961 tests run, 5 skipped, 1 failed.
--none--

**Comment by asfgit** [comment]
SUCCESS 
 8069 tests run, 5 skipped, 0 failed.
--none--

**Comment by mjsax** [comment]
Test failures unrelated. Retest this please.

**Comment by mjsax** [comment]
Retest this please.

**Comment by mjsax** [comment]
Retest this please.

File: clients/src/main/java/org/apache/kafka/clients/CommonClientConfigs.java [hunk 1]
```diff
@@ -59,6 +59,9 @@
     public static final String RECONNECT_BACKOFF_MAX_MS_CONFIG = "reconnect.backoff.max.ms";
     public static final String RECONNECT_BACKOFF_MAX_MS_DOC = "The maximum amount of time in milliseconds to wait when reconnecting to a broker that has repeatedly failed to connect. If provided, the backoff per host will increase exponentially for each consecutive connection failure, up to this maximum. After calculating the backoff increase, 20% random jitter is added to avoid connection storms.";
 
+    public static final String RETRIES_CONFIG = "retries";
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/CommonClientConfigs.java:62]: `AdminClientConfig.RETRIES_CONFIG` and its doc can also piggy back on this value.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyContextImpl.java [hunk 2]
```diff
@@ -67,11 +67,10 @@ public void close() {}
     };
 
     StandbyContextImpl(final TaskId id,
-                       final String applicationId,
                        final StreamsConfig config,
                        final ProcessorStateManager stateMgr,
                        final StreamsMetrics metrics) {
-        super(id, applicationId, config, metrics, stateMgr, new ThreadCache(new LogContext("zeroCache "), 0, metrics));
+        super(id, config, metrics, stateMgr, new ThreadCache(new LogContext("zeroCache "), 0, metrics));
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyContextImpl.java]: This is not part of the PR: though we do not expect it to log any message, but in case of error trouble shooting, should we use a more meaningful log prefix, e.g. `String.format("stream-thread [%s] ", Thread.currentThread().getName())`?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java [hunk 3]
```diff
@@ -68,14 +69,16 @@
      * @throws ProcessorStateException if the base state directory or application state directory does not exist
      *                                 and could not be created
      */
-    public StateDirectory(final String applicationId, final String stateDirConfig, final Time time) {
+    public StateDirectory(final StreamsConfig config,
+                          final Time time) {
         this.time = time;
+        final String stateDirConfig = config.getString(StreamsConfig.STATE_DIR_CONFIG);
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java]: nit: rename to `stateDirName`?

**Comment by asfgit** [comment]
SUCCESS 
 8070 tests run, 5 skipped, 0 failed.
--none--

**Comment by asfgit** [comment]
SUCCESS 
 8070 tests run, 5 skipped, 0 failed.
--none--

**Comment by asfgit** [comment]
SUCCESS 
 8070 tests run, 5 skipped, 0 failed.
--none--

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java [hunk 4]
```diff
@@ -63,23 +65,30 @@
...
+    private final long retryBackoffMs;
 
-    public GlobalStateManagerImpl(final ProcessorTopology topology,
+    public GlobalStateManagerImpl(final LogContext logContext,
+                                  final ProcessorTopology topology,
                                   final Consumer<byte[], byte[]> globalConsumer,
                                   final StateDirectory stateDirectory,
-                                  final StateRestoreListener stateRestoreListener) {
+                                  final StateRestoreListener stateRestoreListener,
+                                  final StreamsConfig config) {
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java:76]: I'd prefer to pass in the two config params here rather than the actual `StreamsConfig` we don't need the entire config.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java [hunk 5]
```diff
@@ -136,7 +145,29 @@ public void register(final StateStore store,
...
-        final Map<TopicPartition, Long> highWatermarks = globalConsumer.endOffsets(topicPartitions);
+        Map<TopicPartition, Long> highWatermarks = null;
+
+        int attempts = 0;
+        while (highWatermarks == null) {
+            try {
+                highWatermarks = globalConsumer.endOffsets(topicPartitions);
+            } catch (final TimeoutException retryableException) {
+                if (++attempts > retries) {
+                    log.error("Failed to get end offsets for topic partitions of global store {} after " + retries + " retry attempts. " +
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java]: nit: use `{}` instead of string concat for `retries`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java [hunk 6]
```diff
@@ -136,7 +145,29 @@ public void register(final StateStore store,
...
+                        "You can increase the number of retries via configuration parameter `retries`.",
+                        store.name(),
+                        retryableException);
+                    throw new StreamsException(String.format("Failed to get end offsets for topic partitions of global store %s after " + retries + " retry attempts. " +
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java]: The string in the log and the exception are identical, extract into into a variable?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java [hunk 7]
```diff
@@ -148,7 +179,30 @@ public void register(final StateStore store,
 
     private List<TopicPartition> topicPartitionsForStore(final StateStore store) {
         final String sourceTopic = topology.storeToChangelogTopic().get(store.name());
-        final List<PartitionInfo> partitionInfos = globalConsumer.partitionsFor(sourceTopic);
+        List<PartitionInfo> partitionInfos = null;
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java]: `null` is redundant

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java [hunk 8]
```diff
@@ -148,7 +179,30 @@ public void register(final StateStore store,
...
+        int attempts = 0;
+        while (true) {
+            try {
+                partitionInfos = globalConsumer.partitionsFor(sourceTopic);
+                break;
+            } catch (final TimeoutException retryableException) {
+                if (++attempts > retries) {
+                    log.error("Failed to get partitions for topic {} after " + retries + " retry attempts. " +
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java]: as above messages in log and exception are identical. Also do we need to log the error with a stacktrace and throw the exception?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStreamThread.java [hunk 9]
```diff
@@ -323,10 +326,16 @@ private StateConsumer initialize() {
...
             stateConsumer.initialize();
             return stateConsumer;
-        } catch (final StreamsException e) {
-            startupException = e;
-        } catch (final Exception e) {
-            startupException = new StreamsException("Exception caught during initialization of GlobalStreamThread", e);
+        } catch (final LockException fatalException) {
+            log.error("Could not lock global state directory. This could happen if multiple KafkaStreams instances " +
+                "are running on the same host using the same state directory.", fatalException);
+            startupException = new StreamsException("Could not lock global state directory. " +
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStreamThread.java]: identical strings in log and exception

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java [hunk 10]
```diff
@@ -68,14 +69,16 @@
      * @throws ProcessorStateException if the base state directory or application state directory does not exist
      *                                 and could not be created
      */
-    public StateDirectory(final String applicationId, final String stateDirConfig, final Time time) {
+    public StateDirectory(final StreamsConfig config,
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java:72]: I'd prefer this was left as it was. The `StateDirectory` doesn't need a `StreamsConfig`.

**Review by dguy — COMMENTED** [review]
Thanks @mjsax, left some comments

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStreamThread.java [hunk 11]
```diff
@@ -323,10 +326,16 @@ private StateConsumer initialize() {
                                         config.getLong(StreamsConfig.COMMIT_INTERVAL_MS_CONFIG));
             stateConsumer.initialize();
             return stateConsumer;
-        } catch (final StreamsException e) {
-            startupException = e;
-        } catch (final Exception e) {
-            startupException = new StreamsException("Exception caught during initialization of GlobalStreamThread", e);
+        } catch (final LockException fatalException) {
+            log.error("Could not lock global state directory. This could happen if multiple KafkaStreams instances " +
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStreamThread.java]: nit: log.fatal?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java (hunk 8 above)
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java]: `... retry attempts due to timeout. The broker may be transiently unavailable at the moment. ..`

Ditto above.

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java [hunk 12]
```diff
@@ -480,7 +476,7 @@ public void onChange(final KafkaStreams.State newState, final KafkaStreams.State
                 }
             }
         });
-        final String appDir = props.getProperty(StreamsConfig.STATE_DIR_CONFIG) + File.separator + props.getProperty(StreamsConfig.APPLICATION_ID_CONFIG);
+        final String appDir = props.get(StreamsConfig.STATE_DIR_CONFIG) + File.separator + props.get(StreamsConfig.APPLICATION_ID_CONFIG);
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java]: Why change to `get` which is less safe since it returns an `Object`?

**Review by guozhangwang — COMMENTED** [review]
Left a few minor comments.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java (hunk 6 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java]: It may be a bit tricky as they use different string concatenation methods.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java (hunk 4 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java:76]: I prefer to keep parameter lists short... Seems to be a matter of taste... Note, we hit style check error in the past if parameter list are too long. Can change it if you insist on it of course. \cc @guozhangwang @bbejeck WDYT?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java (hunk 8 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java]: I thought, for errors we log the whole stack trace -- we don't know if users register an uncaught exception handler, and from my understanding, the exception plus stack trace could get lost otherwise.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java (hunk 10 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java:72]: Same argument as above.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStreamThread.java (hunk 11 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStreamThread.java]: There is no `log.fatal`

**Comment by asfgit** [comment]
SUCCESS 
 8083 tests run, 5 skipped, 0 failed.
--none--

**Comment by asfgit** [comment]
SUCCESS 
 8083 tests run, 5 skipped, 0 failed.
--none--

**Comment by asfgit** [comment]
SUCCESS 
 8083 tests run, 5 skipped, 0 failed.
--none--

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java (hunk 4 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java:76]: I agree it is kind of a personal taste. Personally I prefer one `config` object than multiple ones.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStreamThread.java (hunk 11 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStreamThread.java]: you are right :P

**Comment by guozhangwang** [comment]
LGTM.
